File: src_python/ai_factory/pytorch/hull_white/bermudan_swaptions.py
"""Batched exact-transition Hull-White Bermudan swaption Monte Carlo."""

from __future__ import annotations

from time import perf_counter
from typing import Any

import torch

from ai_factory.pytorch.common.bermudan_lsm import present_value_cashflows
from ai_factory.pytorch.common.device import resolve_device, synchronize, warmup
from ai_factory.pytorch.common.fixed_income import (
    hull_white_bond_coefficients,
    hull_white_integral_variance,
    hull_white_state_integral_covariance,
    hull_white_state_variance,
    nelson_siegel_discount,
)
from ai_factory.pytorch.common.monte_carlo import price_summary

DEFAULT_CPU_BATCH_ROWS = 32
DEFAULT_CUDA_BATCH_ROWS = 128
MAX_EXERCISES = 8
MAX_PAYMENTS = 20
BASIS_RATE_SCALE = 0.04


def _column(values: list[float], target: torch.device) -> torch.Tensor:
    return torch.tensor(values, device=target, dtype=torch.float64)[:, None]


def _price_batch_impl(rows, models, curves, products, *, num_paths, target, batch_rows):
    started = perf_counter()
    outputs: list[dict[str, float]] = []
    payment_axis = torch.arange(MAX_PAYMENTS, device=target)[None, :]
    for offset in range(0, len(rows), batch_rows):
        batch = rows[offset : offset + batch_rows]
        ms = [models[row["model_id"]] for row in batch]
        cs = [curves[row["curve_id"]] for row in batch]
        ps = [products[row["product_id"]] for row in batch]
        batch_size = len(batch)
        a = _column([m["mean_reversion"] for m in ms], target)
        sigma = _column([m["volatility"] for m in ms], target)
        beta0 = _column([c["beta0"] for c in cs], target)
        beta1 = _column([c["beta1"] for c in cs], target)
        beta2 = _column([c["beta2"] for c in cs], target)
        tau = _column([c["tau"] for c in cs], target)
        first = _column([p["first_exercise"] for p in ps], target)
        period = _column([p["exercise_period"] for p in ps], target)
        accrual = _column([p["accrual_period"] for p in ps], target)
        strike = _column([p["fixed_rate"] for p in ps], target)
        notional = _column([p["notional"] for p in ps], target)
        direction = _column([p["direction"] for p in ps], target)
        exercise_count = torch.tensor(
            [p["exercise_count"] for p in ps], device=target, dtype=torch.long
        )
        payment_count = torch.tensor(
            [p["payment_count"] for p in ps], device=target, dtype=torch.long
        )
        exercise_axis = torch.arange(MAX_EXERCISES, device=target)[None, :]
        exercise_times = first + period * exercise_axis
        exercise_active = exercise_axis < exercise_count[:, None]

        bond_a = torch.zeros(
            (batch_size, MAX_EXERCISES, MAX_PAYMENTS),
            device=target,
            dtype=torch.float64,
        )
        bond_b = torch.zeros_like(bond_a)
        for exercise in range(MAX_EXERCISES):
            maturities = first + accrual * (payment_axis + 1)
            coefficients = hull_white_bond_coefficients(
                exercise_times[:, exercise : exercise + 1],
                maturities,
                a,
                sigma,
                beta0,
                beta1,
                beta2,
                tau,
            )
            bond_a[:, exercise, :] = coefficients[0]
            bond_b[:, exercise, :] = coefficients[1]

        state = torch.zeros((batch_size, num_paths), device=target, dtype=torch.float64)
        stochastic_integral = torch.zeros_like(state)
        states = torch.zeros(
            (batch_size, num_paths, MAX_EXERCISES), device=target, dtype=torch.float64
        )
        discounts = torch.ones_like(states)
        previous_time = torch.zeros_like(first)
        for exercise in range(MAX_EXERCISES):
            interval = exercise_times[:, exercise : exercise + 1] - previous_time
            normals = torch.randn(
                (batch_size, num_paths, 2), device=target, dtype=torch.float64
            )
            decay = torch.exp(-a * interval)
            state_variance = hull_white_state_variance(a, sigma, interval)
            integral_variance = hull_white_integral_variance(a, sigma, interval)
            covariance = hull_white_state_integral_covariance(a, sigma, interval)
            state_scale = torch.sqrt(state_variance)
            loading = covariance / state_scale
            previous_state = state
            state = decay * state + state_scale * normals[:, :, 0]
            stochastic_integral = (
                stochastic_integral
                + (-torch.expm1(-a * interval) / a) * previous_state
                + loading * normals[:, :, 0]
                + torch.sqrt(torch.clamp(integral_variance - loading.square(), min=0.0))
                * normals[:, :, 1]
            )
            deterministic_integral = (
                -torch.log(
                    nelson_siegel_discount(
                        exercise_times[:, exercise : exercise + 1],
                        beta0,
                        beta1,
                        beta2,
                        tau,
                    )
                )
                + 0.5
                * hull_white_integral_variance(
                    a, sigma, exercise_times[:, exercise : exercise + 1]
                )
            )
            states[:, :, exercise] = state
            discounts[:, :, exercise] = torch.exp(
                -deterministic_integral - stochastic_integral
            )
            previous_time = exercise_times[:, exercise : exercise + 1]

        immediate = torch.zeros_like(states)
        basis_state = torch.zeros_like(states)
        for exercise in range(MAX_EXERCISES):
            bonds = bond_a[:, exercise, :, None] * torch.exp(
                -bond_b[:, exercise, :, None] * states[:, None, :, exercise]
            )
            active_payment = (
                (payment_axis >= exercise) & (payment_axis < payment_count[:, None])
            )[:, :, None]
            annuity = (active_payment * accrual[:, :, None] * bonds).sum(dim=1)
            end_index = (payment_count - 1)[:, None, None].expand(-1, 1, num_paths)
            end_bond = torch.gather(bonds, 1, end_index).squeeze(1)
            signed_swap = direction * (1.0 - end_bond - strike * annuity)
            immediate[:, :, exercise] = notional * torch.clamp(signed_swap, min=0.0)
            basis_state[:, :, exercise] = (1.0 - end_bond) / annuity / BASIS_RATE_SCALE
        immediate *= exercise_active[:, None, :]
        cashflows = present_value_cashflows(
            immediate, basis_state, discounts, exercise_count
        )
        outputs.extend(price_summary(cashflows))
    synchronize(target)
    return outputs, {"wall_seconds": perf_counter() - started}


def price_batch(rows, model_by_id, curve_by_id, product_by_id, *, num_paths, device,
                batch_rows=None, **_):
    target = resolve_device(device)
    if batch_rows is None:
        batch_rows = (
            DEFAULT_CUDA_BATCH_ROWS
            if target.type == "cuda"
            else DEFAULT_CPU_BATCH_ROWS
        )
    if target.type == "cuda" and rows:
        warmup(target, shape=(min(batch_rows, len(rows)), num_paths, 2), dtype=torch.float64)
        _price_batch_impl(
            rows[:batch_rows], model_by_id, curve_by_id, product_by_id,
            num_paths=num_paths, target=target, batch_rows=batch_rows,
        )
    return _price_batch_impl(
        rows, model_by_id, curve_by_id, product_by_id,
        num_paths=num_paths, target=target, batch_rows=batch_rows,
    )

File: src_cpp/ai_factory/cpu/hull_white/bermudan_swaptions.cpp
#include "ai_factory/cpu/hull_white/bermudan_swaptions.hpp"

#include "ai_factory/common/fixed_income/hull_white.hpp"
#include "ai_factory/cpu/common/bermudan_lsm.hpp"
#include "ai_factory/cpu/common/philox.hpp"

#include <algorithm>
#include <array>
#include <cmath>
#include <stdexcept>
#include <vector>

namespace ai_factory::cpu::hull_white {
namespace {

constexpr int kMaxExercises = 8;
constexpr int kMaxPayments = 20;
constexpr double kBasisRateScale = 0.04;

void validate(const cuda::BermudanSwaptionTerms& product) {
    if (product.exercise_count < 2 || product.exercise_count > kMaxExercises
        || product.payment_count < product.exercise_count + 2
        || product.payment_count > kMaxPayments
        || std::abs(product.exercise_period - product.accrual_period) > 1.0e-12) {
        throw std::invalid_argument("Invalid Bermudan swaption schedule.");
    }
}

}  // namespace

void price_bermudan_swaption_batch(
    const cuda::HullWhiteBermudanSwaptionRow* rows,
    std::size_t row_count,
    std::size_t num_paths,
    cuda::MonteCarloOutput* outputs
) {
    if (num_paths < 2U) {
        throw std::invalid_argument("Bermudan swaption pricing requires two paths.");
    }
#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
    for (std::ptrdiff_t row_index = 0;
         row_index < static_cast<std::ptrdiff_t>(row_count);
         ++row_index) {
        const auto& row = rows[row_index];
        const auto& product = row.product;
        validate(product);
        const auto exercise_count = static_cast<std::size_t>(
            product.exercise_count
        );

        std::array<std::array<double, kMaxPayments>, kMaxExercises> bond_a{};
        std::array<std::array<double, kMaxPayments>, kMaxExercises> bond_b{};
        std::array<double, kMaxExercises> exercise_times{};
        std::array<double, kMaxExercises> decay{};
        std::array<double, kMaxExercises> state_scale{};
        std::array<double, kMaxExercises> integral_loading{};
        std::array<double, kMaxExercises> residual_scale{};
        std::array<double, kMaxExercises> integral_state_loading{};
        std::array<double, kMaxExercises> deterministic_integral{};

        double previous_time = 0.0;
        for (std::size_t exercise = 0; exercise < exercise_count; ++exercise) {
            const double time = product.first_exercise
                                + exercise * product.exercise_period;
            exercise_times[exercise] = time;
            const double interval = time - previous_time;
            decay[exercise] = std::exp(-row.mean_reversion * interval);
            state_scale[exercise] = std::sqrt(
                fixed_income::hull_white_state_variance(
                    row.mean_reversion, row.volatility, interval
                )
            );
            const double integral_variance =
                fixed_income::hull_white_integral_variance(
                    row.mean_reversion, row.volatility, interval
                );
            const double covariance =
                fixed_income::hull_white_state_integral_covariance(
                    row.mean_reversion, row.volatility, interval
                );
            integral_loading[exercise] = covariance / state_scale[exercise];
            residual_scale[exercise] = std::sqrt(std::max(
                integral_variance
                    - integral_loading[exercise] * integral_loading[exercise],
                0.0
            ));
            integral_state_loading[exercise] = fixed_income::hull_white_b(
                row.mean_reversion, interval
            );
            deterministic_integral[exercise] =
                fixed_income::hull_white_deterministic_integral(
                    time,
                    row.mean_reversion,
                    row.volatility,
                    row.beta0,
                    row.beta1,
                    row.beta2,
                    row.tau
                );
            for (int payment = static_cast<int>(exercise);
                 payment < product.payment_count;
                 ++payment) {
                const double maturity = product.first_exercise
                                        + (payment + 1)
                                              * product.accrual_period;
                bond_a[exercise][payment] = fixed_income::hull_white_bond_a(
                    time,
                    maturity,
                    row.mean_reversion,
                    row.volatility,
                    row.beta0,
                    row.beta1,
                    row.beta2,
                    row.tau
                );
                bond_b[exercise][payment] = fixed_income::hull_white_b(
                    row.mean_reversion, maturity - time
                );
            }
            previous_time = time;
        }

        std::vector<double> states(exercise_count * num_paths);
        std::vector<double> discounts(exercise_count * num_paths);
        for (std::size_t path = 0; path < num_paths; ++path) {
            simulation::PhiloxNormalSequence normals(
                row.seed, 0U, path * 2U * exercise_count
            );
            double state = 0.0;
            double stochastic_integral = 0.0;
            for (std::size_t exercise = 0;
                 exercise < exercise_count;
                 ++exercise) {
                const double previous_state = state;
                const double first_normal = normals.next();
                const double second_normal = normals.next();
                state = decay[exercise] * previous_state
                        + state_scale[exercise] * first_normal;
                stochastic_integral +=
                    integral_state_loading[exercise] * previous_state
                    + integral_loading[exercise] * first_normal
                    + residual_scale[exercise] * second_normal;
                const std::size_t index = exercise * num_paths + path;
                states[index] = state;
                discounts[index] = std::exp(
                    -deterministic_integral[exercise] - stochastic_integral
                );
            }
        }

        const auto evaluate = [&](std::size_t exercise, double state) {
            double annuity = 0.0;
            for (int payment = static_cast<int>(exercise);
                 payment < product.payment_count;
                 ++payment) {
                annuity += product.accrual_period * bond_a[exercise][payment]
                           * std::exp(-bond_b[exercise][payment] * state);
            }
            const int last_payment = product.payment_count - 1;
            const double end_bond = bond_a[exercise][last_payment]
                                    * std::exp(
                                        -bond_b[exercise][last_payment] * state
                                    );
            const double signed_swap = static_cast<double>(product.direction)
                                       * (1.0 - end_bond
                                          - product.fixed_rate * annuity);
            const double par_rate = (1.0 - end_bond) / annuity;
            return lsm::ExerciseValue{
                product.notional * std::max(signed_swap, 0.0),
                par_rate / kBasisRateScale,
            };
        };
        const auto cashflows = lsm::bermudan_cashflows_present_value(
            states, discounts, exercise_count, num_paths, evaluate
        );
        double sum = 0.0;
        double sumsq = 0.0;
        for (const double cashflow : cashflows) {
            sum += cashflow;
            sumsq += cashflow * cashflow;
        }
        const double path_count = static_cast<double>(num_paths);
        const double mean = sum / path_count;
        const double variance =
            (sumsq - path_count * mean * mean) / (path_count - 1.0);
        outputs[row_index] = {
            mean,
            std::sqrt(std::max(variance, 0.0) / path_count),
        };
    }
}

}  // namespace ai_factory::cpu::hull_white
